### ISSUE TITLE
xruns at low load

### ISSUE THREAD

**Issue by riban-bw** [zynthian/zynthian-issue-tracking#806]
**Describe the bug**
With no audio sounding and fairly light engines loaded there are frequent instances of xruns.

**To Reproduce**
Steps to reproduce the behaviour:
1. Load FluidR3 GM snapshot
2. Add a MIDI plugin in series with one of the chains, e.g. MIDI Strum

**Expected behaviour**
System runs without xruns.

**Actual behaviour**
Xruns every second with approx. 9ms delay.

**Setup:**
- HARDWARE
Raspberry Pi 4 Model B Rev 1.4
Audio: [HifiBerry DAC+ ADC PRO](http://192.168.1.51/hw-audio)
Display: [ZynScreen 3.5 (v1)](http://192.168.1.51/hw-display)
Wiring: [MCP23017_ZynScreen](http://192.168.1.51/hw-wiring)
I2C: [MCP23017@0x20, MCP23008@0x21, ADS1115@0x48](http://192.168.1.51/hw-wiring)
- SYSTEM
Raspbian GNU/Linux 10 (buster)
Build Date: 2022-10-28
Memory: 19% (361M/1872M)
SD Card: 40% (12G/29G)
Temperature: 53.5ºC
Overclock: [None](http://192.168.1.51/hw-options)
- MIDI
Profile: [default.sh](http://192.168.1.51/ui-midi-options)
Tuning: [440 Hz](http://192.168.1.51/ui-midi-options)
Receive Mode: [Stage (Omni On)](http://192.168.1.51/ui-midi-options)
ZS3 Sub-SnapShots: [On](http://192.168.1.51/ui-midi-options)
MIDI to Output: [Off](http://192.168.1.51/ui-midi-options)
Master Channel: [off](http://192.168.1.51/ui-midi-options)
- SOFTWARE
zyncoder: [staging-2210 (98c3df2)](https://github.com/zynthian/zyncoder/commit/98c3df25193485902ecdd2da4a312e5a3c87843b)
zynthian-ui: [staging-2210 (9a85e14)](https://github.com/zynthian/zynthian-ui/commit/9a85e14a784adcedba8f112e27698ee68715bac1)
zynthian-sys: [staging-2210 (8c7e2ac)](https://github.com/zynthian/zynthian-sys/commit/8c7e2ac6b8a1209f73f2319f74a326d75bd03d8d)
zynthian-data: [staging-2210 (3bbc39c)](https://github.com/zynthian/zynthian-data/commit/3bbc39cb3febe968dd2be3155f547db1848dba71)
zynthian-webconf: [staging-2210 (b60b529)](https://github.com/zynthian/zynthian-webconf/commit/b60b5292d056a3966ddf105a6399ef3bd85bdba7)

**Additional context**
Although this snapshot loads 16 chains it is a single engine with low resource usage. `jack_cpu_load` shows approx. 2% and `top` shows all 4 cores running below 3%. 1GB RAM is free.

**Comment by riban-bw** [comment]
Removing the MIDI plugin does not stop the xruns.
Clean All does not stop the xruns.
Eventually see this in the log:
```
Oct 29 06:41:32 zynthian startx[2550]: WARNING:zynthian_autoconnect.cb_jack_xrun: Jack Audio XRUN! => delayed 3463.0us
Oct 29 06:41:33 zynthian startx[2550]: SuspendRefNum error
Oct 29 06:41:33 zynthian startx[2550]: JackClient::Execute error name = zynsmf
```
After this Jack was not running.

**Comment by riban-bw** [comment]
I wonder if this may be caused by something else. The previous day I had played the test audio and test MIDI then left the device powered overnight. The xruns were observed when I started to use it the following morning, approx. 9 hours later. The jack error may indicate an issue with zynsmf?

**Comment by riban-bw** [comment]
The jack error occured whilst SMF library was idle (no SMF being played) which indicates some background task. It is plausible that jack changed its reported samplerate (maybe due to some other issue like xrun) which could have been reported as zero. That could have triggered a divide by zero bug in smf library (which is fixed by commit [9d682a0](https://github.com/zynthian/zynthian-ui/commit/9d682a0b6cbc982f9f2a9178356ff0c688e9f147)). That might explain the jack/smf error but I don't think it was the cause of the xruns. I think excessive xruns triggered the smf error.

So still need to figure out what caused Zynthian to start xrun every second. I did not notice xruns before adding the plugins.

**Comment by riban-bw** [comment]
I suspect this is related to the problem that closing zynthian does not always close the engines and modules gracefully. A restart can begin starting modules before they have fully closed which can result in unexpected state and bad performance. We need a mechanism to ensure modules are closed properly with checks that this has happened. I have started to work on some improvements to the module and service management in "service improvements" branch. This is a work in progress that needs some careful consideration of dependencies, module interaction, etc. I am fairly confident that this issue is not as described in the title and that most normal workflows should not hit this issue, i.e. users are not restarting zynthian ui so I will close the ticket and progress the core issues described here as a stability development.
